### ISSUE TITLE
python39Packages.pytest-xdist is non-deterministic

### ISSUE THREAD

**Issue by trofi** [NixOS/nixpkgs#139292]
### Describe the bug

My ulteriour goal is o track down why python-xdist's tests fail occasionally for me. A step in that direction is to try to rebuild a package locally a few times. Unfortunately the package is not deterministic:

```
$ nix build -f . python39Packages.pytest-xdist --rebuild
error: derivation '/nix/store/vxlbh98pfhzvwpyq1xqmd1wgdj0zk6c9-python3.9-pytest-xdist-2.3.0.drv' may not be deterministic: output '/nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0' differs
```

### Notify maintainers

@dotlambda 

### Metadata

```console
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.14.3, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210908_3c56f62`
 - channels(root): `"nixos-21.11pre317242.bc9b956714e"`
 - channels(slyfox): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: python39Packages.pytest-xdist
```

**Comment by trofi** [comment]
The only diff is in some pycache files `.../python3.9/site-packages/xdist/__pycache__/__init__.cpython-39-pytest-6.2.5.pyc`:

```diff
$ diffoscope '/nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0' '/nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0.check'
--- /nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0
+++ /nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0.check
│   --- /nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0/lib
├── +++ /nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0.check/lib
│ │   --- /nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0/lib/python3.9
│ ├── +++ /nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0.check/lib/python3.9
│ │ │   --- /nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0/lib/python3.9/site-packages
│ │ ├── +++ /nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0.check/lib/python3.9/site-packages
│ │ │ │   --- /nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0/lib/python3.9/site-packages/xdist
│ │ │ ├── +++ /nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0.check/lib/python3.9/site-packages/xdist
│ │ │ │ │   --- /nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0/lib/python3.9/site-packages/xdist/__pycache__
│ │ │ │ ├── +++ /nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0.check/lib/python3.9/site-packages/xdist/__pycache__
│ │ │ │ │ │   --- /nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0/lib/python3.9/site-packages/xdist/__pycache__/__init__.cpython-39-pytest-6.2.5.pyc
│ │ │ │ │ ├── +++ /nix/store/vgdj386fvjdk3fjjpb70swibmly2r38f-python3.9-pytest-xdist-2.3.0.check/lib/python3.9/site-packages/xdist/__pycache__/__init__.cpython-39-pytest-6.2.5.pyc
│ │ │ │ │ │ @@ -1,8 +1,8 @@
│ │ │ │ │ │ -00000000: 610d 0d0a 0000 0000 bd10 3561 3101 0000  a.........5a1...
│ │ │ │ │ │ +00000000: 610d 0d0a 0000 0000 ae81 4d61 3101 0000  a.........Ma1...
│ │ │ │ │ │  00000010: e300 0000 0000 0000 0000 0000 0000 0000  ................
│ │ │ │ │ │  00000020: 0002 0000 0040 0000 0073 4a00 0000 6400  .....@...sJ...d.
│ │ │ │ │ │  00000030: 6401 6c00 5a01 6400 6401 6c02 6d03 0200  d.l.Z.d.d.l.m...
│ │ │ │ │ │  00000040: 0100 6d04 5a05 0100 6400 6402 6c06 6d07  ..m.Z...d.d.l.m.
│ │ │ │ │ │  00000050: 5a07 6d08 5a08 6d09 5a09 6d0a 5a0a 0100  Z.m.Z.m.Z.m.Z...
│ │ │ │ │ │  00000060: 6400 6403 6c0b 6d0c 5a0d 0100 6700 6404  d.d.l.m.Z...g.d.
│ │ │ │ │ │  00000070: a201 5a0e 6401 5300 2905 e900 0000 004e  ..Z.d.S.)......N
```

What are they? Could it an effect of missing propagated `pytest` inputs (and thus cache is built against different version of pytest)?

**Comment by trofi** [comment]
Looks like the only change is timestamp difference:

```
--- /dev/fd/63  2021-09-24 09:46:15.642812835 +0100
+++ /dev/fd/62  2021-09-24 09:46:15.642812835 +0100
@@ -1,5 +1,5 @@
 magic b'610d0d0a'
-moddate b'bd103561' (Sun Sep  5 19:47:25 2021)
+moddate b'ae814d61' (Fri Sep 24 08:43:42 2021)
```

Python-3.9 should already honor `SOURCE_DATE_EPOCH`. I wonder why it does not work here.

**Comment by trofi** [comment]
Might be an effect of `pytest-xdist-2.3.0/setup.py: use_scm_version={"write_to": "src/xdist/_version.py"},`? I see other python packages in `nixpkgs` sed it out:

```
pkgs/development/python-modules/FormEncode/default.nix:    sed -i 's/use_scm_version=.*,/version="${version}",/' setup.py
```

**Comment by veprbl** [comment]
Might want to check out #22570

**Comment by trofi** [comment]
> Might want to check out #22570

Oh, nice! It has some hints about manual bytecode generation. I'll take a closer look.

Meanwhile filed request against `diffoscope` to get better default rendering of `.pyc` diffs: https://salsa.debian.org/reproducible-builds/diffoscope/-/issues/278

**Comment by trofi** [comment]
Aha, now I get it a bit better: `pytest` only mimics `python` in how it writes bytecode (for better assertion handling), but does not exactly follow it. The hint is that deviations always happen in `.*-pytest-.*.pyc` files.

`pytest` generates it's own headers in https://github.com/pytest-dev/pytest/blob/2aaea20cb219e7d66751f51508ba2a5f80c4c009/src/_pytest/assertion/rewrite.py#L279-L295

Which is: `u32(MAGIC) | u32(0) | u32(mtime) | u32(size)`. `mtime` is an unstable bit here.

`cpython` generates headers in https://github.com/python/cpython/blob/2f180ce2cb6e6a7e3c517495e0f4873d6aaf5f2f/Lib/importlib/_bootstrap_external.py#L694-L712

Which is (assuming `SOURCE_DATE_EPOCH` is set): `u32(MAGIC) | u32(0x3) | u64(source_hash) | u32(size)`. No unstable bits here.

**Comment by trofi** [comment]
I noticed that `pytest` itself works it around by wiping out `pytest` bytecode for it not being stable: https://github.com/NixOS/nixpkgs/blob/6bb29572e4dc9410401bd6b6d166549be8427c78/pkgs/development/python-modules/pytest/default.nix#L71-L72

Is it desirable to install those file? (suppose we fix determinism for them) Or it's always a sign of unexpected leftovers?

Casual scan on my local system shows quite a few `-pytest-.*.pyc` files:

```
$ find /nix/store -name '*-pytest-*.pyc' | unnix | tr '/' ' ' | awk '{print $2}' | uniq | sort -u | uniq
python3.9-aiohttp-3.7.4.post0
python3.9-async_generator-1.10
python3.9-Bottleneck-1.3.2
python3.9-debugpy-1.4.3
python3.9-et_xmlfile-1.0.1
python3.9-flaky-3.7.0
python3.9-hypothesis-6.17.3
python3.9-nbval-0.9.6
python3.9-numpy-1.21.2
python3.9-pandas-1.3.2
python3.9-pytest-6.1.2
python3.9-pytest-flake8-1.0.7
python3.9-pytest-forked-1.3.0
python3.9-pytest-mock-3.6.1
python3.9-pytest-randomly-3.10.1
python3.9-pytest-timeout-1.4.2
python3.9-pytest-xdist-2.3.0
python3.9-pytest-xdist-2.3.0.check
python3.9-pyzmq-22.2.1
python3.9-scipy-1.7.1
python3.9-SQLAlchemy-1.4.22
python3.9-SQLAlchemy-1.4.23
python3.9-SQLAlchemy-1.4.25
python3.9-tables-3.6.1
python3.9-typeguard-2.12.1
```

**Comment by trofi** [comment]
Proposed bytecode cleanup as https://github.com/NixOS/nixpkgs/pull/140179.

It does not fix all the packages above as some of them bypass pytest hook and run `py.test` directly. We can either port them to a hook or do the same manual cleanup.

**Comment by trofi** [comment]
Was fixed by #140179
